src/Raven.Server/Documents/Sharding/Background/ShardedDocumentsMigrator.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Server.ServerWide.Commands.Sharding;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Logging;
+
+namespace Raven.Server.Documents.Sharding.Background
+{
+    public class ShardedDocumentsMigrator : IDisposable
+    {
+        private readonly ShardedDocumentDatabase _database;
+        private readonly CancellationTokenSource _cts;
+        private readonly Logger _logger;
+
+        public ShardedDocumentsMigrator(ShardedDocumentDatabase database)
+        {
+            _database = database;
+            _cts = CancellationTokenSource.CreateLinkedTokenSource(database.ShardedDocumentsStorage.DocumentDatabase.DatabaseShutdown);
+            _logger = LoggingSource.Instance.GetLogger(database.Name, GetType().FullName);
+        }
+
+        internal async Task ExecuteMoveDocumentsAsync()
+        {
+            _cts.Token.ThrowIfCancellationRequested();
+
+            if (Monitor.TryEnter(this, 250) == false)
+                return;
+
+            if (_database.ServerStore.Sharding.HasActiveMigrations(_database.ShardedDatabaseName))
+                return;
```

Comment:
`Monitor.Exit`?

Reply:
note that this might still be racy and you will miss a cleanup.

1. (thread 1) You are here with the locked taken and a migration is ongoing
2. (thread 2) One moment later the migration is done and we enter this method again but you are blocked by the Monitor and return at this code `Monitor.TryEnter(this, 250)` - nothing happened
3. (thread 1) return here and nothing happened as well

Do you recover form this situation?

Reply:
In general, why do we need the `Monitor` here, is it possible to avoid it?

Reply:
Done